### ISSUE TITLE
Generative tests are consistently failing with a Trino connection error

### ISSUE THREAD

**Issue by evansd** [opensafely-core/ehrql#1567]
They run for 2-3 hours and then fail with an enormous stack of connection errors like:
```
Traceback (most recent call last):
  File "/home/runner/work/ehrql/ehrql/tests/generative/test_query_model.py", line 326, in run_with
    engine.setup(instances, metadata=sqla_metadata)
  File "/home/runner/work/ehrql/ehrql/tests/conftest.py", line 177, in setup
    return self.database.setup(*items, metadata=metadata)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/tests/lib/databases.py", line 107, in setup
    metadata.create_all(engine)
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/sql/schema.py", line 5792, in create_all
    bind._run_ddl_visitor(
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 3239, in _run_ddl_visitor
    conn._run_ddl_visitor(visitorcallable, element, **kwargs)
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2443, in _run_ddl_visitor
    visitorcallable(self.dialect, self, **kwargs).traverse_single(element)
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py", line 670, in traverse_single
    return meth(obj, **kw)
           ^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/sql/ddl.py", line 901, in visit_metadata
    [t for t in tables if self._can_create_table(t)]
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/sql/ddl.py", line 901, in <listcomp>
    [t for t in tables if self._can_create_table(t)]
                          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/sql/ddl.py", line 866, in _can_create_table
    return not self.checkfirst or not self.dialect.has_table(
                                      ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/trino/sqlalchemy/dialect.py", line 335, in has_table
    res = connection.execute(sql.text(query), {"schema": schema, "table": table_name})
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1412, in execute
    return meth(
           ^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 483, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1635, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1844, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1984, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2342, in _handle_dbapi_exception
    raise exc_info[1].with_traceback(exc_info[2])
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/trino/sqlalchemy/dialect.py", line 373, in do_execute
    cursor.execute(statement, parameters)
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/trino/dbapi.py", line 590, in execute
    self._iterator = iter(self._query.execute())
                          ^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/trino/client.py", line 806, in execute
    response = self._request.post(self._query, additional_http_headers)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/trino/client.py", line 555, in post
    http_response = self._post(
                    ^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/trino/client.py", line 899, in decorated
    raise error
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/trino/client.py", line 886, in decorated
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/requests/sessions.py", line 637, in post
    return self.request("POST", url, data=data, json=json, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/ehrql/ehrql/.venv/lib/python3.11/site-packages/requests/adapters.py", line 519, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='localhost', port=32769): Max retries exceeded with url: /v1/statement (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f240551b710>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

Full example [here](https://github.com/opensafely-core/ehrql/actions/runs/6079812550/job/16492833128) although that link will expire eventually.

This is despite the waving of several dead chickens in the hope that the error might go away:
 * https://github.com/opensafely-core/ehrql/pull/1552
 * https://github.com/opensafely-core/ehrql/pull/1561

I think the first step is to try to get the logs from the Trino container. I think probably the easiest way of doing this would be to add something to the test session teardown (maybe use the `recorder` fixture?) which shells out to Docker to grab the logs. The full log session is likely to be enormous so probably we only want the last N lines for some reasonable N.

You can make these changes on a branch and then use the manual workflow dispatch button to run the tests against your branch. This has the added benefit that failures won't cause a notification in the Slack channel.

It might be possible to shorten the feedback loop here by, somewhere inside the `run_test()` function, catching the connection error and then immediately grabbing the Trino logs and calling `sys.exit()` rather than letting Hypothesis keep trying examples against a dead instance.

It's possible that the logs will make it obvious what's going on, but it's equally possible they'll tell us nothing useful at all.

Trino (and the JVM itself) has various knobs and levers we can try fiddling with:
https://trino.io/docs/current/installation/deployment.html#configuring-trino
https://trino.io/docs/current/admin/properties.html

All the relevant config files are available in:
[`tests/support/trino/etc`](https://github.com/opensafely-core/ehrql/tree/522d9ef8b28cf58878362083fac583d376bbc0d7/tests/support/trino/etc)

**Comment by evansd** [comment]
I tried to get some more debugging information by running against [this commit](https://github.com/opensafely-core/ehrql/commit/27de2b1095a6a8f982eba822a7670ba3d6d99110) which I fondly imagined might dump the logs of the Trino container on exit.

Of course what it actually [logged](https://github.com/opensafely-core/ehrql/actions/runs/6253311069/job/16978424478) was:
```
Error response from daemon: No such container: ehrql-trino
```

Which I probably could have predicted if I'd thought about it a bit harder. I wonder if this is because the container is getting OOM killed?

In any case, I could try streaming the logs to a file and then grabbing them from the file on exit.

**Comment by evansd** [comment]
Finally got some [debug info](https://github.com/opensafely-core/ehrql/actions/runs/6258518604/job/16992679168) out using [this commit](https://github.com/opensafely-core/ehrql/commit/80ca32aa8e677eec03a57a0c5c95fbac72ef2670). As suspected, it's a memory issue:
```
java.lang.OutOfMemoryError: Java heap space
Dumping heap to java_pid1.hprof ...
Error: [7671.202s][error][jvmti] Posting Resource Exhausted event: Java heap space
ResourceExhausted: Java heap space: killing current process!
```

The full log shows it struggling with GC prior to this:
[`trino-logs.txt`](https://github.com/opensafely-core/ehrql/files/12683308/trino-logs.txt)

I'm going to try to reproduce this locally by constraining the memory available to the Docker image.

Then I'll try fiddling with the [JVM config](https://github.com/opensafely-core/ehrql/blob/2b7b236f952270718118fe78ecbe34a829a50724/tests/support/trino/etc/jvm.config) to see if we can prevent this happening. There's no reason it should need loads of memory to run a single query, it's obviously just not releasing it fast enough.

**Comment by evansd** [comment]
It's easy to reproduce this locally if you restrict the Docker container to, say, 1.5GB.

The issue seems to be either (a) there's a memory leak or (b) the garbage collector just can't keep up with the objects being created.

I wondered if the issue might be related to our use of the memory connector. But fiddling with the code so that we only did one set of INSERTs and then ran all the queries against the same data didn't seem to make any difference so I don't think it's that.

I went a short way down the rabbit hole of fiddling with JVM GC config options before coming to my senses.

I think we just need a mechanism for restarting the database; either the entire container or maybe just the process within the container, which turns out to be structurally easiest.
